containers/org.eclipse.linuxtools.docker.ui/src/org/eclipse/linuxtools/internal/docker/ui/jobs/CopyFromDockerJob.java

```diff
@@ -0,0 +1,1233 @@
+		} finally {
+			m_monitor.done();
+		}
+		return Status.OK_STATUS;
+	}
+
+	/**
+	 * Copy a folders or files from the image to the host. If symlinks point outside
+	 * the source folder, they are resolved and the actual files are copied up to a
+	 * depth of MAXLINKDEPTH.
+	 *
+	 * @return Ok, or Cancel if the job was canceled by the user. If something went
+	 *         wrong Error will be returned.
+	 */
+	private IStatus runCopy() {
+
+		Path currentVolume = null;
+
+		// TODO image is not set if container
+		m_monitor.beginTask(Messages.getFormattedString(COPY_VOLUMES_FROM_DESC, m_image), m_pathsToCopy.size() + 2);
+
+		try (final CloseableContainer container = (m_containerId == null ? new CloseableContainer(m_connection, m_image)
+				: null); Closeable dcToken = m_connection.getOperationToken()) {
+
+			if (container != null) {
+				m_containerId = container.containerId;
+				container.start();
+			}
+
+			// copy each volume if it exists and is not copied over yet
+			for (Entry<Path, Path> pathToCopy : m_pathsToCopy.entrySet()) {
+
+				Path srcdir = pathToCopy.getKey();
+				Path hostDir = pathToCopy.getValue();
+
+				currentVolume = srcdir;
+
+				m_monitor.worked(1);
+				// TODO better message
+				copyFolderFromContainer(dcToken, srcdir, (Path) hostDir.append(srcdir.lastSegment()), 0);
+			}
+		} catch (OperationCanceledException e) {
+			return Status.CANCEL_STATUS;
+		} catch (DockerException | InterruptedException e) {
+			Activator.logErrorMessage("Docker Connection Error: " + e.getMessage(), e);
+			//TODO Set messge
+		} catch (Exception e) {
+			Activator.logErrorMessage("Failed to copy a from b" + currentVolume + " " + m_image, e);
+			// TODO Set messge
+		} finally {
+			m_monitor.done();
+		}
+		return Status.OK_STATUS;
+	}
+
+	/**
+	 * Copy a folder
+	 *
+	 * @param dcToken   The closable docker connection Token
+	 * @param srcdir    The path in the container
+	 * @param hostDir   The path on the host
+	 * @param linkdepth The current link depth
+	 * @throws DockerException
+	 * @throws InterruptedException
+	 * @throws IOException
+	 */
+	private void copyFolderFromContainer(Closeable dcToken, Path srcdir, Path hostDir, int linkdepth)
+			throws DockerException, InterruptedException, IOException {
+
+		if (m_mirror) {
+			throw new RuntimeException("This must not be called in mirror-mode");
+		}
+
+		if (m_monitor.isCanceled()) {
+			return;
+		}
+
+		m_monitor.subTask(Messages.getFormattedString(COPY_VOLUMES_FROM_TASK, srcdir.toString()));
+
+		Set<SymLink> symlinkBacklog = new HashSet<>();
+
+		try (InputStream in = m_connection.copyContainer(dcToken, m_containerId, srcdir.toString())) {
+			copyTar(srcdir, hostDir, in, symlinkBacklog, m_monitor);
+		}
+
+		for (SymLink sl : symlinkBacklog) {
+			if (!sl.targetInBase()) {
+				// The link goes outside what was just copied
+				if (linkdepth < MAXLINKDEPTH) {
+					copyFolderFromContainer(dcToken, sl.m_target, sl.hostPath(), linkdepth + 1);

```

Comment:
The linkdepth never gets incremented anywhere.  Did you mean to have it be: ++linkdepth instead of linkdepth + 1?

Reply:
? `linkdepth` is a local variable that is incremented with every recursive call. `++linkdepth` would be wrong as it would increment the local recursion level.

Reply:
Ah, ok.